devDocs/startupShutdownInstall.md

```diff
@@ -0,0 +1,82 @@
+    1. NVDA+q
+        - test: `startupShutdownNVDA.Quits from keyboard, Restarts`
+    1. An input gesture to restart
+    1. A prompt to restart from changing certain settings
+    1. Via the NVDA menu -> Exit
+        - test: `startupShutdownNVDA.Quits from menu`
+1. A process sending `WM_QUIT`, eg a new NVDA process starting (uses `WM_QUIT`)
+1. A handled crash (directly causes a new process to start, terminates unsafely)
+    - test: `startupShutdownNVDA.Restarts on crash, Restarts on braille crash`
+1. An unhandled crash (terminates unsafely)
+1. An external command which kills the process (terminates unsafely) 
+1. Windows shutting down (terminates unsafely) (uses `wx.EVT_END_SESSION`)
+
+## Exit hooks/triggers
+
+### In process: `triggerNVDAExit`
+* A function in the core module
+* Only executes the code once, uses a lock and flag to ensure this
+* Uses a queue on the main thread to queue a safe shutdown
+* Once the queued shutdown starts:
+    1. the updateCheck is terminated
+    1. watchdog is terminated
+    1. addons and the brailleViewer are terminated, so we can close all windows safely
+    1. All NVDA windows are closed
+    1. Now that windows are closed, a new NVDA instance is started if requested
+
+### Out of process: `WM_QUIT`
+* [a windows message](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-quit) which can be sent across processes which will force the main loop to exit and close most NVDA windows.
+* We subsequently run `triggerNVDAExit` to ensure that clean up code isn't missed, and pump the queue to execute it.
+* Changing to using a custom message, which would allow a custom handling (eg just `triggerNVDAExit`), requires compatibility with messaging older NVDA versions that are only aware of `WM_QUIT`.
+
+### Out of process: `wx.EVT_END_SESSION`
+- This is a [wxCloseEvent](https://docs.wxwidgets.org/3.0/classwx_close_event.html) triggered by a Windows session ending.
+- On `wx.EVT_END_SESSION`, we save the config and play the exit sound.
+- Other actions are not performed as we have limited time to perform an action for this event. With NVDA being killed off very last, Windows may include NVDA in the Block shutdown dialog, but the user won't be able to read it if we are shutting down.
+
+## Technical startup and exit notes
+
+## The Message Window
+
+The Windows API allows us to find a process by searching for a named window. To make the NVDA process searchable, NVDA creates a message window with a specific name.
+
+Starting up is blocked by checking for an existing NVDA window and sending `WM_QUIT`.
+We then set the NVDA process to allow the `WM_QUIT` to be sent from other processes (see [WM_QUIT](#Out-of-process-WM_QUIT)). 
+
+The message window is created late during the start up, and destroyed early in exit and is not perfectly indicative of whether or not an NVDA process is running. As such, we have a [MutEx](#MutEx) that is checked as well.
+

```

Comment:
I think the intent of the message window / window class lookup, is so that that `WM_QUIT` can be sent. Otherwise we can't identify the process, right? This only matters if NVDA isn't already exiting. If the message window doesn't exist, it is either too early, or NVDA is already exiting.

I'd try to make this intent clear:
- The WM_QUIT / window class mechanism is to make it possible to trigger another NVDA process to START to exit.
- The MutEx ensures that a newly started process blocks until any previous NVDA has FINISHED exiting.

Reply:
I've rewritten this to make the purpose of the MW clearer. The Message Window is used for handling other windows messages (see `MessageWindow`), not just `WM_QUIT`. But in regards to this document, that is why it is relevant.